In [1]:
# ExtraTreesRegressor
# sklearn
csv_path = r"D:\DSpit\AIH\lab\Diabetic-Patients-Readmission-Prediction\CSV Files\Diabetes_cleaned.csv"

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv(csv_path)

In [4]:
df.head()

race  gender  age discharge_disposition_id admission_source_id  \
0        Caucasian       0    5            Not Available            Referral   
1        Caucasian       0   15       Discharged to home           Emergency   
2  AfricanAmerican       0   25       Discharged to home           Emergency   
3        Caucasian       1   35       Discharged to home           Emergency   
4        Caucasian       1   45       Discharged to home           Emergency   

   time_in_hospital  num_lab_procedures  num_procedures  num_medications  \
0                 1                  41               0                1   
1                 3                  59               0               18   
2                 2                  11               5               13   
3                 2                  44               1               16   
4                 1                  51               0                8   

                                        diag_1  ... glipizide glyburide  \
0                                     Diabetes  ...        -2        -2   
1  Endocrine, Nutritional, Metabolic, Immunity  ...        -2        -2   
2                        Pregnancy, Childbirth  ...         0        -2   
3                     Infectious and Parasitic  ...        -2        -2   
4                                    Neoplasms  ...         0        -2   

   pioglitazone rosiglitazone insulin  diabetesMed  readmitted  \
0            -2            -2      -2            0           0   
1            -2            -2       1            1           0   
2            -2            -2      -2            1           0   
3            -2            -2       1            1           0   
4            -2            -2       0            1           0   

   preceding_year_visits  number_changes  insulin_treatment  
0                      0             0.0             no_med  
1                      0             1.0       insulin_only  
2                      3             0.0         other_meds  
3                      0             1.0       insulin_only  
4                      0             0.0      insulin_combo  

[5 rows x 28 columns]

In [5]:
df.columns

Index(['race', 'gender', 'age', 'discharge_disposition_id',
       'admission_source_id', 'time_in_hospital', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'glimepiride', 'glipizide', 'glyburide', 'pioglitazone',
       'rosiglitazone', 'insulin', 'diabetesMed', 'readmitted',
       'preceding_year_visits', 'number_changes', 'insulin_treatment'],
      dtype='object')

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97070 entries, 0 to 97069
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   race                      97070 non-null  object 
 1   gender                    97070 non-null  int64  
 2   age                       97070 non-null  int64  
 3   discharge_disposition_id  97070 non-null  object 
 4   admission_source_id       97070 non-null  object 
 5   time_in_hospital          97070 non-null  int64  
 6   num_lab_procedures        97070 non-null  int64  
 7   num_procedures            97070 non-null  int64  
 8   num_medications           97070 non-null  int64  
 9   diag_1                    97070 non-null  object 
 10  diag_2                    97070 non-null  object 
 11  diag_3                    97070 non-null  object 
 12  number_diagnoses          97070 non-null  int64  
 13  max_glu_serum             5092 non-null   object 
 14  A1Cres

In [6]:
# Index(['race', 'gender', 'age', 'discharge_disposition_id',
#        'admission_source_id', 'time_in_hospital', 'num_lab_procedures',
#        'num_procedures', 'num_medications', 'diag_1', 'diag_2', 'diag_3',
#        'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
#        'repaglinide', 'glimepiride', 'glipizide', 'glyburide', 'pioglitazone',
#        'rosiglitazone', 'insulin', 'diabetesMed', 'readmitted',
#        'preceding_year_visits', 'number_changes', 'insulin_treatment'],
#       dtype='object')



Data preprocessing is an essential step in preparing a dataset for machine learning models. Here are the detailed and advanced data preprocessing steps for a DataFrame with the given columns:

1. Handling Missing Values:
   - Identify columns with missing values using `df.isnull().sum()`. In this case, `max_glu_serum` and `A1Cresult` have missing values.
   - Decide on a strategy to handle missing values. Options include:
     - Dropping rows or columns with missing values using `df.dropna()`.
     - Imputing missing values with statistical measures like mean, median, or mode using `df.fillna()`.
     - Using advanced techniques like regression or K-nearest neighbors to impute missing values.

2. Handling Categorical Variables:
   - Identify columns with categorical variables. In this case, `race`, `discharge_disposition_id`, `admission_source_id`, `diag_1`, `diag_2`, `diag_3`, `max_glu_serum`, `A1Cresult`, and `insulin_treatment` are categorical.
   - Encode categorical variables using techniques like one-hot encoding or label encoding. One-hot encoding creates binary columns for each category, while label encoding assigns a numerical label to each category.

3. Feature Scaling:
   - Check if any numerical columns require feature scaling. In this case, `age`, `num_lab_procedures`, `num_procedures`, `num_medications`, `number_diagnoses`, `preceding_year_visits`, `number_changes` are numerical.
   - Apply feature scaling techniques like standardization or normalization to ensure all features have a similar scale. Standardization scales the data to have zero mean and unit variance, while normalization scales the data to a specific range.

4. Handling Imbalanced Data (if applicable):
   - Check if the target variable (`time_in_hospital`) is imbalanced, meaning some classes have significantly more instances than others.
   - If imbalanced, consider techniques like oversampling (e.g., SMOTE) or undersampling to balance the classes and prevent bias towards the majority class.

5. Feature Engineering:
   - Explore the existing features and create new features that might be more informative for the model.
   - For example, you could create new features by combining or transforming existing features, such as calculating the BMI from height and weight or creating interaction terms.

6. Removing Outliers (if applicable):
   - Identify potential outliers in numerical columns using techniques like box plots or z-scores.
   - Decide on a strategy to handle outliers, such as removing them or transforming them using techniques like winsorization.

7. Splitting the Data:
   - Split the dataset into training, validation, and testing sets using techniques like train-test split or cross-validation.
   - This allows you to train the model on the training set, tune hyperparameters on the validation set, and evaluate the final model on the testing set.

These are some of the detailed and advanced data preprocessing steps you can perform on the given DataFrame. The specific steps and techniques used may vary depending on the characteristics of the dataset and the requirements of the machine learning task.

In [21]:
target = "time_in_hospital"

In [23]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.metrics import r2_score
from sklearn.impute import SimpleImputer


Fitting 5 folds for each of 100 candidates, totalling 500 fits


c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:547: FitFailedWarning: 
160 fits failed out of a total of 500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
137 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 1474, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\pipeline.py", line 475, in fit
    self._final_estima

Best R-squared on validation set: 0.8602
R-squared on test set: 0.8584

Best hyperparameters:
regressor__n_estimators: 165
regressor__min_samples_split: 18
regressor__min_samples_leaf: 1
regressor__max_features: sqrt
regressor__max_depth: 70


In [ ]:

# Define column types
categorical_columns = ['race', 'discharge_disposition_id', 'admission_source_id', 'diag_1', 'diag_2', 'diag_3', 'max_glu_serum', 'A1Cresult', 'insulin_treatment']
numerical_columns = ['gender', 'age', 'time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_diagnoses', 'metformin', 'repaglinide', 'glimepiride', 'glipizide', 'glyburide', 'pioglitazone', 'rosiglitazone', 'insulin', 'diabetesMed', 'readmitted', 'preceding_year_visits']

target = 'number_changes'  # Assuming this is your target variable

# Create preprocessing steps
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numerical_columns),
        ('cat', categorical_transformer, categorical_columns)
    ])


In [29]:

# Create pipeline
pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('regressor', ExtraTreesRegressor(
        n_estimators=165,
        min_samples_split=18,
        min_samples_leaf=1,
        max_features='sqrt',
        max_depth=70,
        random_state=42
    ))
])


In [ ]:

# Split the data
X = df.drop(columns=[target])
y = df[target]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:

# Define parameter distribution for RandomizedSearchCV
param_dist = {
    'regressor__n_estimators': np.arange(50, 500),
    'regressor__max_depth': [None] + list(np.arange(10, 110, 10)),
    'regressor__min_samples_split': np.arange(2, 21),
    'regressor__min_samples_leaf': np.arange(1, 21),
    'regressor__max_features': ['auto', 'sqrt', 'log2']
}

# Perform RandomizedSearchCV
random_search = RandomizedSearchCV(pipeline, param_distributions=param_dist, 
                                   n_iter=100, cv=5, scoring='r2', n_jobs=-1, 
                                   random_state=42, verbose=1)
random_search.fit(X_train, y_train)

# Get best model and score
best_model = random_search.best_estimator_
best_score = random_search.best_score_

# Evaluate on test set
y_pred = best_model.predict(X_test)
test_r2 = r2_score(y_test, y_pred)

print(f"Best R-squared on validation set: {best_score:.4f}")
print(f"R-squared on test set: {test_r2:.4f}")
print("\nBest hyperparameters:")
for param, value in random_search.best_params_.items():
    print(f"{param}: {value}")

Fitting 5 folds for each of 100 candidates, totalling 500 fits
c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:547: FitFailedWarning: 
160 fits failed out of a total of 500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
137 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 1474, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\pipeline.py", line 475, in fit
    self._final_estimator.fit(Xt, y, **last_step_params["fit"])
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 1467, in wrapper
    estimator._validate_params()
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 666, in _validate_params
    validate_parameter_constraints(
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\utils\_param_validation.py", line 95, in validate_parameter_constraints
    raise InvalidParameterError(
sklearn.utils._param_validation.InvalidParameterError: The 'max_features' parameter of ExtraTreesRegressor must be an int in the range [1, inf), a float in the range (0.0, 1.0], a str among {'sqrt', 'log2'} or None. Got 'auto' instead.

--------------------------------------------------------------------------------
23 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 1474, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\pipeline.py", line 475, in fit
    self._final_estimator.fit(Xt, y, **last_step_params["fit"])
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 1467, in wrapper
    estimator._validate_params()
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\base.py", line 666, in _validate_params
    validate_parameter_constraints(
  File "c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\utils\_param_validation.py", line 95, in validate_parameter_constraints
    raise InvalidParameterError(
sklearn.utils._param_validation.InvalidParameterError: The 'max_features' parameter of ExtraTreesRegressor must be an int in the range [1, inf), a float in the range (0.0, 1.0], a str among {'log2', 'sqrt'} or None. Got 'auto' instead.

  warnings.warn(some_fits_failed_message, FitFailedWarning)
c:\Users\Vedant\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:1051: UserWarning: One or more of the test scores are non-finite: [       nan 0.82337029 0.74324546 0.84841261        nan        nan
 0.60941673 0.75918705 0.68912592        nan 0.62480436 0.69972251
 0.62095102 0.79274427 0.63334765 0.76711552 0.55068741        nan
 0.7576993  0.61081725        nan 0.73905605 0.41482245 0.82287758
 0.60888521 0.79014696        nan 0.79742831 0.55986459        nan
 0.82216734        nan        nan 0.63201955 0.6991526  0.64846847
        nan 0.6734847         nan        nan        nan 0.68707801
        nan 0.73201365 0.74924208 0.74444351 0.79706712 0.57776299
        nan 0.76105077 0.85023714 0.73897626 0.60539909 0.5998205
        nan 0.75823048 0.67387096 0.60212276 0.84953593 0.73351961
        nan 0.58285818 0.78675087 0.6689912  0.77826018 0.73998352
 0.61843473 0.7587924  0.77503952 0.57571271        nan 0.78269309
 0.82314528        nan 0.74934345 0.86022671        nan 0.59203234
        nan        nan        nan 0.74301368 0.60094924 0.73158772
 0.40159866        nan 0.75998187        nan        nan        nan
        nan 0.63805262 0.65906188 0.7923661         nan 0.74127332
        nan 0.59384814        nan 0.74763294]
  warnings.warn(
Best R-squared on validation set: 0.8602
R-squared on test set: 0.8584

Best hyperparameters:
regressor__n_estimators: 165
regressor__min_samples_split: 18
regressor__min_samples_leaf: 1
regressor__max_features: sqrt
regressor__max_depth: 70

In [7]:
# sklearn.ensemble.ExtraTreesRegressor
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error


In [24]:
X_train.head()

race  gender  age                 discharge_disposition_id  \
51499        Caucasian       0   85  Transferred to another medical facility   
32493  AfricanAmerican       0   75                       Discharged to home   
53949  AfricanAmerican       0   65                       Discharged to home   
78710  AfricanAmerican       1   45                       Discharged to home   
78688            Asian       0   65                       Discharged to home   

      admission_source_id  time_in_hospital  num_lab_procedures  \
51499           Emergency                 4                  72   
32493            Referral                 7                  50   
53949           Emergency                 2                  52   
78710           Emergency                 3                  47   
78688           Emergency                 5                  66   

       num_procedures  num_medications  \
51499               0                8   
32493               3               51   
53949               0               20   
78710               1               23   
78688               3               10   

                                             diag_1  ... glimepiride  \
51499  Musculoskeletal System and Connective Tissue  ...          -2   
32493                                   Circulatory  ...          -2   
53949                                   Circulatory  ...           0   
78710                                   Respiratory  ...          -2   
78688                                   Circulatory  ...          -2   

      glipizide  glyburide pioglitazone rosiglitazone  insulin  diabetesMed  \
51499        -2         -2           -2            -2       -2            0   
32493         0         -2           -2            -2        0            1   
53949        -2         -2           -2            -2        0            1   
78710        -2         -2           -2            -2       -1            1   
78688        -2         -2           -2            -2       -2            0   

       readmitted  preceding_year_visits  insulin_treatment  
51499           0                      0             no_med  
32493           0                      0      insulin_combo  
53949           0                      1      insulin_combo  
78710           0                      1       insulin_only  
78688           0                      0             no_med  

[5 rows x 27 columns]

In [30]:
y_train.head()

51499    0.0
32493    0.0
53949    0.0
78710    1.0
78688    0.0
Name: number_changes, dtype: float64

In [31]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['gender', 'age',
                                                   'time_in_hospital',
                                                   'num_lab_procedures',
                                                   'num_procedures',
                                                   'num_medications',
                                                   'number_diagnoses',
                                                   'metformin', 'repaglinide',
                                                   'glimepiride', 'glipizide',
                                                   'glyburide', 'pioglitazone'...
                                                                   SimpleImputer(fill_value='missing',
                                                                                 strategy='constant')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['race',
                                                   'discharge_disposition_id',
                                                   'admission_source_id',
                                                   'diag_1', 'diag_2', 'diag_3',
                                                   'max_glu_serum', 'A1Cresult',
                                                   'insulin_treatment'])])),
                ('regressor',
                 ExtraTreesRegressor(max_depth=70, max_features='sqrt',
                                     min_samples_split=18, n_estimators=165,
                                     random_state=42))])

In [32]:
y_pred = pipeline.predict(X_test)

In [33]:
mse = mean_squared_error(y_test, y_pred)

In [34]:
r2 = r2_score(y_test, y_pred)

In [35]:
print(f"Mean Squared Error: {mse}")
print(f"R-squared Score: {r2}")

Mean Squared Error: 0.0334162881376403
R-squared Score: 0.858386878286313


In [36]:
# Feature importance
feature_importance = pipeline.named_steps['regressor'].feature_importances_
feature_names = (pipeline.named_steps['preprocessor']
                 .named_transformers_['num'].get_feature_names_out(numerical_columns).tolist() + 
                 pipeline.named_steps['preprocessor']
                 .named_transformers_['cat'].get_feature_names_out(categorical_columns).tolist())

# Create a DataFrame of feature importances
feature_importance_df = pd.DataFrame({'feature': feature_names, 'importance': feature_importance})
feature_importance_df = feature_importance_df.sort_values('importance', ascending=False)

print("\nTop 10 Most Important Features:")
print(feature_importance_df.head(10))


Top 10 Most Important Features:
                             feature  importance
14                           insulin    0.514447
102         insulin_treatment_no_med    0.051890
103     insulin_treatment_other_meds    0.045054
7                          metformin    0.042641
15                       diabetesMed    0.038324
100  insulin_treatment_insulin_combo    0.034873
101   insulin_treatment_insulin_only    0.034027
11                         glyburide    0.031616
10                         glipizide    0.028577
9                        glimepiride    0.011772
